File: Programmers/DFS,BFS/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> num;
int tar, answer;

void go(int index, int sum)
{
    if(index == num.size())
    {
        if(sum == tar)
            answer++;
        return;
    }
    go(index + 1, sum + num[index]); 
    go(index + 1, sum - num[index]);
    
}

int solution(vector<int> numbers, int target) 
{
    num = numbers;
    tar = target;
    go(0,0);
    return answer;
}

File: C&Python/DFS/타겟넘버.py
answer = 0
def dfs(numbers, target, cur_total, index):
    global answer
    #exit condition
    if index == len(numbers):
        if cur_total == target:
            answer += 1
        return
    
    # +
    dfs(numbers, target, cur_total + numbers[index], index + 1)
    
    # -
    dfs(numbers, target, cur_total - numbers[index], index + 1)
    
def solution(numbers, target):
    dfs(numbers, target, 0, 0)
    return answer